### ISSUE TITLE
Unhandled promise rejection on redirect

### ISSUE THREAD

**Issue by rosskevin** [4Catalyzer/found#210]
https://github.com/4Catalyzer/found#redirects

We successfully use redirects elsewhere in our app, but are encountering a problem with our dashboard redirect - for some reason I can't spot.  I've searched and cannot find a related issue. 

Here is a snippet of code from the route:

```jsx
      <Route
        path="/"
        Component={loadable(() => import(/* webpackChunkName: "Landing" */ './LandingPage'))}
        render={args => {
          const { Component, props } = args
          const subdomain = resolveSubdomain()
          if (subdomain) {
            throw new RedirectException('/dashboard')
          } else {
            return !Component ? undefined : <Component {...props} />
          }
        }}
      />
```

The results of this are:

```
Unhandled promise rejection 
Object { location: "/dashboard" }
es6.promise.js:110
onUnhandled/</result<
es6.promise.js:110
kiCJ/module.exports
_perform.js:3:12
onUnhandled/<
es6.promise.js:104
[6725]/module.exports
_invoke.js:5
<anonymous>
_task.js:35
run
_task.js:21
listener
_task.js:25
```

For context, this route is rendered within the following:

```jsx
const RootRouteComponent = (props: AppFrameProps) => (
  <TenantSwitcher>
    <GoogleAnalyticsProvider>
      <AppFrame {...props} />
    </GoogleAnalyticsProvider>
  </TenantSwitcher>
)


      <Route>
        <Route Component={RootRouteComponent}>
          {this.resolve(LandingRoutes)} {/* Landing routes resolves to the code above */}
          {/* We do something similar here in renderAuthenticate and it works fine */}
          <Route render={this.renderAuthenticate}>
            {this.resolve(AccessRoutes)}
          </Route>
        </Route>
      </Route>
```

The routes guarded by `<Route render={this.renderAuthenticate}>` serve conditional redirects fine, the only thing different is that the redirect logic is on a parent route, whereas in the dashboard redirect, it is on the route with the component itself.

What am I missing here?

**Comment by taion** [comment]
Are you using Found Relay here? If so, can you check what value `resolving` on the render args has, per https://github.com/relay-tools/found-relay#render?

**Comment by rosskevin** [comment]
@taion not using found-relay.

I did just trace a path though, and if not authenticated, there is a potential for two redirects - with that said, even authenticated (single redirect) isn't working for this landing vs dashboard page.

Potential paths:
- / redirect -> /dashboard redirect -> signin
- / redirect -> /dashboard

**Comment by rosskevin** [comment]
I've taken away the `loadable` temporarily and no change.

**Comment by taion** [comment]
It wouldn't be anything promise-related with the component. Could something outside of the Found resolver be calling `render` on the route?

What seems a little odd to me is that `render` should be getting called via `createElements` here, but I'm not seeing that in your stack trace. There shouldn't be any additional async there.

**Comment by rosskevin** [comment]
I've confirmed it is being called from `createElements`:

![firefox_developer_edition_and_unhandled_promise_rejection_on_redirect_ _issue__210_ _4catalyzer_found](https://user-images.githubusercontent.com/136564/47674669-daa24100-db85-11e8-8242-eea55ff7bac6.png)

**Comment by rosskevin** [comment]
We have a base class called `RouteConfig` which has functions reused with other contextual setup classes/state.  @bmac151 just said moving the render function to the base class worked, but in the subclass it fails.  Investigating that now.

**Comment by rosskevin** [comment]
Confirmed, so these functions don't work like this, but moving to the base class allows them to work.

Fails:

```tsx
import { RedirectException, RouteRenderArgs } from 'found'
import { Route } from 'found'
import * as React from 'react'

import DashboardPage from './Dashboard/DashboardPage'
import LandingPage from './Landing/LandingPage'

const log = Logger.get('AppRouteConfig')

const RootRouteComponent = (props: AppFrameProps) => (
  <TenantSwitcher>
    <GoogleAnalyticsProvider>
      <AppFrame {...props} />
    </GoogleAnalyticsProvider>
  </TenantSwitcher>
)

//        {/* {this.resolve(SamplesRoutes)} */} add these in dev mode?
export default class extends RouteConfig {
  public renderTenantRedirectToDashboard = (args: RouteRenderArgs) => {
    log.debug('renderTenantRedirectToDashboard', args)
    const { Component, props } = args
    const subdomain = resolveSubdomain()
    log.debug('subdomain', subdomain)
    if (subdomain && subdomain !== 'public') {
      log.debug('throw redirect exception')
      throw new RedirectException('/dashboard')
    } else {
      log.debug('normal path')
      return !Component ? undefined : <Component {...props} />
    }
  }

  public debugRender = (args: RouteRenderArgs) => {
    log.debug('render /dashboard route', args)
    const { Component, props } = args
    log.debug('normal path')
    return !Component ? undefined : <Component {...props} />
  }

  public routes() {
    return (
      <Route>
        <Route Component={RootRouteComponent}>
          <Route
            render={this.renderTenantRedirectToDashboard}
            path="/"
            Component={LandingPage}
            marketing
          />
          <Route render={this.debugRender} path="/dashboard" Component={DashboardPage} />
        </Route>
      </Route>
    )
  }
}
```
But when the `renderTenantRedirectToDashboard` and `debugRender` are moved to the base class, it works.

Still trying some things but I'm not sure I see the problem here yet.

**Comment by rosskevin** [comment]
So I'm missing something fundamental with `this`.

The following works in the base `RouteConfig` class but not in the subclass:

```tsx
public debugRender = (args: RouteRenderArgs) => {}
public debugRender(args: RouteRenderArgs) {}
```

It seems like these only work in the base class.  I'm going to inline them (which is what I originally did) and see what happens.

**Comment by rosskevin** [comment]
Inlining does not work.  Even though these render functions are using nothing from the base class, they will not work unless they are in the base class (which is not feasible - the base class is used for many products and this is product specific).

This fails:

```tsx
const log = Logger.get('AppRouteConfig')

const RootRouteComponent = (props: AppFrameProps) => (
  <TenantSwitcher>
    <GoogleAnalyticsProvider>
      <AppFrame {...props} />
    </GoogleAnalyticsProvider>
  </TenantSwitcher>
)

//        {/* {this.resolve(SamplesRoutes)} */} add these in dev mode?
export default class extends RouteConfig {
  public routes() {
    return (
      <Route>
        <Route Component={RootRouteComponent}>
          <Route
            render={(args: RouteRenderArgs) => {
              log.debug('renderTenantRedirectToDashboard', args)
              const { Component, props } = args
              const subdomain = resolveSubdomain()
              log.debug('subdomain', subdomain)
              if (subdomain && subdomain !== 'public') {
                log.debug('throw redirect exception')
                throw new RedirectException('/dashboard')
              } else {
                log.debug('normal path')
                return !Component ? undefined : <Component {...props} />
              }
            }}
            path="/"
            Component={LandingPage}
            marketing
          />
          <Route
            render={(args: RouteRenderArgs) => {
              log.debug('render /dashboard route', args)
              const { Component, props } = args
              log.debug('normal path')
              return !Component ? undefined : <Component {...props} />
            }}
            path="/dashboard"
            Component={DashboardPage}
          />
        </Route>
      </Route>
    )
  }
}
```

**Comment by rosskevin** [comment]
I'm missing something fundamental here - if you know what I'm happy to spend time reading - just point me in a direction.  I can't fathom why inlining or using in the subclass fails.

**Comment by taion** [comment]
I don't really know what's going on here. A couple of possibilities:

- This `RedirectException` is somehow a different class. Maybe drop a breakpoint in https://github.com/4Catalyzer/found/blob/ca594e7d1800ac7d45ab6e82fd1a7b994945c964/src/createBaseRouter.js#L142 and see if that check is failing.
- The compilation does something different in the subclass case... check if the emitted code looks somehow weird

**Comment by rosskevin** [comment]
I have pulled out the base class and just using the inline to reduce complexity.  You are absolutely right about checking the instanceof class.  Most of the `found` setup happens in our common lib, but routes are created in the app.  It is using two different versions of `found` in development, so the `instanceof` check is failing.

I can add a webpack alias - is this the right solution?

**Comment by taion** [comment]
I would do that for now. The right fix is that instead of doing `instanceof`, we should probably add some static `errorType` field on the class or something, and check that.

**Comment by taion** [comment]
@jquense I know you've gotten bitten by this in the context of GraphQL.js before. What do you think is the right way to check for the error type here?

**Comment by hisapy** [comment]
Hey devs, 

As requested by @taion in Slack, I'm posting my experience with a very similar error here, but in my case it was when trying to catch a request error coming from a `Promise.reject`. 

For the record, I'm currently on the following versions:
├─ found-relay@0.4.0-alpha.3
└─ found@0.4.0-alpha.6

Basically, my problem was that the `renderError` of the following router (based on `found-relay`) was never called.

```javascript
const FarceRouter = createFarceRouter({
  historyProtocol: new BrowserProtocol(),
  historyMiddlewares: [queryMiddleware],
  routeConfig: routeConfig,

  render: createRender({
    renderError({ error, router, location }) {
      if (process.env.NODE_ENV !== "test") {
        // eslint-disable-next-line no-console
        console.error(error);
      }

      switch (error.status) {
        case 401:
          return <Login router={router} location={location} />;
        case 404:
          return <h1>Página no encontrada o en construcción</h1>;
        default:
          return <h1>Ha ocurrido un error. Intente recargar la página</h1>;
      }
    }
  })
});
```

At first I thought it was a problem with Relay v2 ... my server was returning a response with status 401 that was `Promise.reject`ed in my Relay Network Layer, but then I realized that `renderError` was not being called even after removing the Relay part and creating a route just to _render_ just a `throw new HttpError(404, "hola")`.

The app was crashing with:
```
asyncToGenerator.js:24 Uncaught (in promise) HttpError {status: 404, data: "hola"}
```

Then I suspected the problem was caused because my `Route` component was imported from an internal package called `auth-required-route` that is added as a dependency in a CRA app, both packages in a Lerna+Yarn workspaces monorepo. Since the module is so simple and small, and because CRA is configured to load _modern_ Javascript by default but not JSX from node_modules, I just added a build step to pass the component through `@babel/plugin-transform-react-jsx`. `found` and `react` are added as `peerDependency` in `auth-required-route`.  So in order to discard this, I copied the code to the CRA, and the `renderError` started working again.

Anyway, I need this `auth-required-route` package to work so I imported again its `AuthRequiredRoute` that was just throwing the `HttpError` and started the debugger. The code didn't stop by the catch, like it wasn't a `HttpError`. In the process I saw some parts going into some babel stuff, so I went to see the package.json of the CRA and it had a `babel-loader`. I don't remember why I installed it because CRA already has all the babel stuff, it even had a different version from the CRA.

Finally, I removed the babel-loader, from the CRA, and cleaned my node_modules and everything started working as expected. I don't what was exactly the solution but I strongly suspect that `babel-loader` or other `babel` stuff was involved in it.
